test/soake2erekt/kafkasource_soak_test.go

```diff
@@ -0,0 +1,334 @@
+			RunSoakFeatureFnWithMapping(ctx, env, t, eventshubReceiverFeature, namesFn)
+			RunSoakFeatureFnWithMapping(ctx, env, t, kafkaSourceScenarioTopicAndSinkSetupFeature, namesFn)
+		},
+		IterationFn: func(ctx context.Context, env environment.Environment, t *testing.T) {
+			RunSoakFeatureFnWithMapping(ctx, env, t, kafkaSourceScenarioInstallKafkaSourceFeature, namesFn)
+			RunSoakFeatureFnWithMapping(ctx, env, t, kafkaSourceScenarioIsReadyKafkaSourceFeature, namesFn)
+			RunSoakFeatureFnWithMapping(ctx, env, t, kafkaSinkSendFeature, namesFn)
+			RunSoakFeatureFnWithMapping(ctx, env, t, verifySingleEventReceivedFeature, namesFn)
+		},
+		TeardownFn: func(ctx context.Context, env environment.Environment, t *testing.T) {
+			f := verifyNoKafkaSourceLeftInDispatcherConfigMap()
+			env.Test(ctx, t, f)
+		},
+	}
+
+	RunSoakTestWithDefaultCopies(t, soakTest)
+}
+
+/*
+TestKafkaSourceAddingAndRemovingSoak on each iteration,
+creates 16 kafkasources from a single topic
+sends an event
+verifies an event is received 16 times
+*/
+func TestKafkaSourceAddingAndRemovingSoak(t *testing.T) {
+	t.Parallel()
+
+	topicPrefix := feature.MakeRandomK8sName("topic")
+	kafkaSinkPrefix := topicPrefix
+	kafkaSourcePrefix := feature.MakeRandomK8sName("kafkaSource")
+	senderPrefix := feature.MakeRandomK8sName("sender")
+
+	// Number of kafkasources to create and delete at once on each iteration
+	const max = 16
+
+	namesFn := func(env SoakEnv) kafkaSourceScenarioNames {
+		return kafkaSourceScenarioNames{
+			receiver:   "receiver",
+			sender:     fmt.Sprintf("%s-%d-%d", senderPrefix, env.CopyID(), env.Iteration()),
+			kafkaTopic: fmt.Sprintf("%s-%d", topicPrefix, env.CopyID()),
+			kafkaSink:  fmt.Sprintf("%s-%d", kafkaSinkPrefix, env.CopyID()),
+			// kafkaSource:  not used, generated below dynamically for each of the 16 kafkasources
+		}
+	}
+
+	soakTest := SoakTest{
+		NamespacePrefix: "test-kafka-source-addrm-",
+		SetupFn: func(ctx context.Context, env environment.Environment, t *testing.T) {
+			RunSoakFeatureFnWithMapping(ctx, env, t, eventshubReceiverFeature, namesFn)
+			RunSoakFeatureFnWithMapping(ctx, env, t, kafkaSourceScenarioTopicAndSinkSetupFeature, namesFn)
+		},
+		// As part of this soak test, we crate 16 kafkasources, then wait for them to be ready,
+		// and finally send an event and verify an event was received 16 times
+		IterationFn: func(ctx context.Context, env environment.Environment, t *testing.T) {
+			for j := 0; j < max; j++ {
+				soakEnv := SoakEnvFromContext(ctx)
+				names := namesFn(soakEnv)
+				names.kafkaSource = fmt.Sprintf("%s-%d-%d", kafkaSourcePrefix, soakEnv.Iteration(), j)
+
+				f := kafkaSourceScenarioInstallKafkaSourceFeature(names)
+				env.Test(ctx, t, f)
+			}
+
+			for j := 0; j < max; j++ {
+				soakEnv := SoakEnvFromContext(ctx)
+				names := namesFn(soakEnv)
+				names.kafkaSource = fmt.Sprintf("%s-%d-%d", kafkaSourcePrefix, soakEnv.Iteration(), j)
+
+				f := kafkaSourceScenarioIsReadyKafkaSourceFeature(names)
```

Comment:
Sounds like kafkaSourceScenarioIsReadyKafkaSourceFeature and kafkaSourceScenarioInstallKafkaSourceFeature should be merged into one. After all, it's one resource, we're installing it and waiting for it to be ready. I don't see any separate usage of "IsReady" that would justify separating them and having more complex code (e.g. the two loops above)

Reply:
we want to first create all, and then wait for all, so that the actual creation can happen in parallel.

Reply:
Hmm, using `env.ParallelTest(ctx, t, f)` should have the same effect. There could be just one feature per KafkaSource, which creates and waits for the kafka source to be ready.

Reply:
Since the tests run with high "-parallel" it could create the KafkaSource all in parallel. The behaviour would be similar but probably better to check if it works as expected.

Reply:
in the overall soak test, this will be just one scenario of many, so I am not sure if we want to do too many parallel request in the single scenario...

Reply:
OK. No problem.